src/state_machine.zig

```diff
@@ -908,6 +887,97 @@ pub fn StateMachineType(comptime Storage: type) type {
         fn get_posted(self: *const StateMachine, pending_id: u128) ?bool {
             return self.forest.grooves.posted.get(pending_id);
         }
+
+        pub fn forest_options(options: Options) Forest.Options {
+            const batch_accounts_max = @intCast(u32, @divFloor(
+                options.message_body_size_max,
+                @sizeOf(Account),
+            ));
+            const batch_transfers_max = @intCast(u32, @divFloor(
+                options.message_body_size_max,
+                @sizeOf(Transfer),
+            ));
+            assert(batch_accounts_max > 0);
+            assert(batch_transfers_max > 0);
+
+            return .{
+                .accounts = .{
+                    .cache_size = options.cache_size_accounts,
+                    .prefetch_count_max = std.math.max(
+                        // create_account()/lookup_account() looks up 1 account per item.
+                        batch_accounts_max,
+                        // create_transfer()/post_or_void_pending_transfer() looks up 2
+                        // accounts for every transfer.
+                        2 * batch_transfers_max,
+                    ),
+                    .tree_options_object = .{
+                        .commit_count_max = math.max(
+                            batch_accounts_max,
+                            // ×2 because creating a transfer will update 2 accounts.
+                            2 * batch_transfers_max,
+                        ),
+                    },
+                    .tree_options_id = .{ .commit_count_max = batch_accounts_max },
+                    .tree_options_index = .{
+                        .user_data = .{ .commit_count_max = batch_accounts_max },
+                        .ledger = .{ .commit_count_max = batch_accounts_max },
+                        .code = .{ .commit_count_max = batch_accounts_max },
+                        // Transfers mutate the secondary indices for debits/credits pending/posted.
+                        //
+                        // * Each mutation results in a remove and an insert: the ×2 multiplier.
+                        // * Each transfer modifies two accounts. However, this does not
+                        //   necessitate an additional ×2 multiplier — the credits of the debit
+                        //   account and the debits of the credit account are not modified.
+                        .debits_pending = .{
+                            .commit_count_max = math.max(
+                                batch_accounts_max,
+                                2 * batch_transfers_max,
+                            ),
+                        },
```

Comment:
This is so powerful. Nice work.

Reply:
Been having fun getting my head around the 2x multiplier for debits/credits the past few days. This is a great explanation.